File: SocketPython/decode.py
from code import stringChecksum, is_num_type
from csv import DATA_LOG, TRANS_LOG

def decodeUpdateString(string):
	if type(string) is type(""):	
		local_checksum = stringChecksum(string[:-4])
		try:
			received_sum = int(string[-4:])
			if received_sum == local_checksum:
				string = string.lower()

				if len(string) == 35:
					if string[0] == "u":
						data = []
						data.append(string[1:9].strip())
						data.append(string[9:17])
						data.append(string[17:32])
						data.append(string[23:31])
						data.append(DATA_LOG)

						if "-" in data[1]:
							data[1] = data[1].replace("-", "", 1)
							data[1] = "-" + data[1]
						for split in data[1:-1]:
							float(split)
						data[2] = data[2][0:2] + ":" + data[2][2:4] + ":" + data[2][4:6]
						data[3] = data[3][0:2] + "-" + data[3][2:4] + "-" + data[3][4:]

						return data
				else:
					print("El mensaje no tiene la longitud correcta ("+str(len(string))+" != 35)")
					return False
			else:
				print("El checksum de la cadena no coincide")
				return False
		except ValueError:
			print("Error en valores numericos")
	else:
		return False

def decodeObserverString(string):
	if type(string) is type(""):	
		local_checksum = stringChecksum(string[:-4])
		try:
			received_sum = int(string[-4:])
			if received_sum == local_checksum:
				string = string.lower()

				if len(string) == 21:
					if string[0] == "r":
						data = []
						data.append(string[1:9].strip())
						data.append(string[9:17].strip())
						data.append(TRANS_LOG)

						return data
				else:
					print("El mensaje no tiene la longitud correcta ("+str(len(string))+" != 35)")
					return False
			else:
				print("El checksum de la cadena no coincide")
				return False
		except ValueError:
			print("Error en valores numericos")
	else:
		return False

def decode_packet(string):
	if string[0] == "r":
		return decodeObserverString(string)
	elif string[0] == "u":
		return decodeUpdateString(string)
	else:
		print("Codificacion desconocida")
		return False

File: SocketRuby/decode.rb
load "code.rb"
load "csv.rb"

$DATA_LOG = "data.csv"
$TRANS_LOG = "transaction_log.csv"

class String
  def numeric?
    Float(self) != nil rescue false
  end
end

def decodeUpdateString(string)
	if string.is_a? String
		local_checksum = stringChecksum(string[0,string.length-4])
		received_sum = string[-4,4].to_i
		if local_checksum == received_sum
			string.downcase!
			if string.length == 35
				if string[0] == "u"
					data = Array.new
					data.append(string[1,8].strip)
					data.append(string[9,8])
					data.append(string[17,6])
					data.append(string[23,8])
					data.append($DATA_LOG)

					if data[1].include? "-"
						data[1].sub!("-","")
						data[1] = "-" + data[1]
					end
					for split in data[1,3]
						if split.numeric? == false
							return false
						end
					end
					data[2] = data[2][0,2] + ":" + data[2][2,2] + ":" +data[2][4,2]
					data[3] = data[3][0,2] + "-" + data[3][2,2] + "-" +data[3][4,4]

					return data
				end
			else
				puts "El mensaje no tiene la longitud correcta #{string.length} != 35"
			end
		else
			puts "El checksum no coincide"
		end
	end
	return false
end

def decodeObserverString(string)
	if string.is_a? String
		local_checksum = stringChecksum(string[0,string.length-4])
		received_sum = string[-4,4].to_i
		if local_checksum == received_sum
			string.downcase!
			if string.length == 21
				if string[0] == "r"
					data = Array.new
					data.append(string[1,8].strip)
					data.append(string[9,8].strip)
					data.append($TRANS_LOG)

					return data
				end
			else
				puts "El mensaje no tiene la longitud correcta #{string.length} != 21"
			end
		else
			puts "El checksum no coincide"
		end
	end
	return false
end

def decode_packet(string)
	if string[0] == "r"
		return decodeObserverString(string)
	elsif string[0] == "u"
		return decodeUpdateString(string)
	else
		puts "Codificacion desconocida\n"
		return false
	end
end
